synapse/event_auth.py

```diff
@@ -80,6 +80,31 @@ def validate_event_for_room_version(event: "EventBase") -> None:
     Raises:
         SynapseError if there is a problem with the event
     """
+    # Reject events with stringy power levels if required by room version
+    if (
+        event.type == EventTypes.PowerLevels
+        and event.room_version.msc3667_int_only_power_levels
+    ):
+        for key, value in event.content.items():
+            if key in [
+                "users_default",
+                "events_default",
+                "state_default",
+                "ban",
+                "reject",
+                "kick",
+                "invite",
+            ]:
+                if not isinstance(value, int):
+                    raise AuthError(403, f"{value} must be an integer.")
+            if key in ["events", "notifications"]:
```

Comment:
I think we need to include `users` too. The [spec wording](https://spec.matrix.org/latest/rooms/v10/#authorization-rules) is:

> 2. **[New in this version]** If either of the keys `events` or `notifications` in content are present and not a dictionary with values that are integers, reject.
> 3. If `users` key in `content` is not a dictionary with keys that are valid user IDs with values that are integers, reject.

and 2. doesn't mention `users` because it's covered in 3. We currently validate a weaker form of 3. in `_check_power_levels`, where we check if the value is parseable as an integer.

Reply:
Thanks for pointing this out, I added users and removed the weaker check in `_check_power_levels` as it is superseded by this change.